src/hotspot/share/opto/memnode.cpp

```diff
@@ -3281,19 +3286,31 @@ Node* MergePrimitiveArrayStores::make_merged_input_value(const Node_List& merge_
     jlong mask = (((jlong)1) << bits_per_store) - 1;
     for (uint i = 0; i < merge_list.size(); i++) {
       jlong con_i = merge_list.at(i)->in(MemNode::ValueIn)->get_int();
+#ifdef VM_LITTLE_ENDIAN
       con = con << bits_per_store;
       con = con | (mask & con_i);
+#else // VM_LITTLE_ENDIAN
+      con_i = (mask & con_i) << (i * bits_per_store);
+      con = con | con_i;
+#endif // VM_LITTLE_ENDIAN
     }
     merged_input_value = _phase->longcon(con);
   } else {
     // Pattern: [base >> 24, base >> 16, base >> 8, base] -> base
     //             |                                  |
     //           _store                             first
     //
-    merged_input_value = first->in(MemNode::ValueIn);
     Node const* base_last;
     jint shift_last;
+#ifdef VM_LITTLE_ENDIAN
+    merged_input_value = first->in(MemNode::ValueIn);
     bool is_true = is_con_RShift(_store->in(MemNode::ValueIn), base_last, shift_last);
+#else // VM_LITTLE_ENDIAN
+    // `_store` points to the lowest using store in the Memory chain. On big endian it stores the
+    //  unshifted `base`. `_store` and `first` need to be exchanged in the diagram above
+    merged_input_value = _store->in(MemNode::ValueIn);
+    bool is_true = is_con_RShift(first->in(MemNode::ValueIn), base_last, shift_last);
+#endif // VM_LITTLE_ENDIAN
```

Comment:
You could just have local variables for "lo" / "hi", set them depending on big/little endian, and then the logic would be the same for both.

Reply:
Yeah, that's better. I've done that.